onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -465,24 +465,31 @@ static void RegisterExecutionProviders(InferenceSession* sess, const std::vector
 #endif
     } else if (type == kCudaExecutionProvider) {
 #ifdef USE_CUDA
-      const auto it = provider_options_map.find(type);
-      CUDAExecutionProviderInfo info{};
-      if (it != provider_options_map.end())
-        GetProviderInfo_CUDA()->CUDAExecutionProviderInfo__FromProviderOptions(it->second, info);
-      else {
-        info.device_id = cuda_device_id;
-        info.gpu_mem_limit = gpu_mem_limit;
-        info.arena_extend_strategy = arena_extend_strategy;
-        info.cudnn_conv_algo_search = cudnn_conv_algo_search;
-        info.do_copy_in_default_stream = do_copy_in_default_stream;
-        info.external_allocator_info = external_allocator_info;
-      }
+      if(auto* cuda_provider_info = TryGetProviderInfo_CUDA())
+      {
+        const auto it = provider_options_map.find(type);
+        CUDAExecutionProviderInfo info{};
+        if (it != provider_options_map.end())
+          cuda_provider_info->CUDAExecutionProviderInfo__FromProviderOptions(it->second, info);
+        else {
+          info.device_id = cuda_device_id;
+          info.gpu_mem_limit = gpu_mem_limit;
+          info.arena_extend_strategy = arena_extend_strategy;
+          info.cudnn_conv_algo_search = cudnn_conv_algo_search;
+          info.do_copy_in_default_stream = do_copy_in_default_stream;
+          info.external_allocator_info = external_allocator_info;
+        }
 
-      // This variable is never initialized because the APIs by which is it should be initialized are deprecated, however they still
-      // exist are are in-use. Neverthless, it is used to return CUDAAllocator, hence we must try to initialize it here if we can
-      // since FromProviderOptions might contain external CUDA allocator.
-      external_allocator_info = info.external_allocator_info;
-      RegisterExecutionProvider(sess, *GetProviderInfo_CUDA()->CreateExecutionProviderFactory(info));
+        // This variable is never initialized because the APIs by which is it should be initialized are deprecated, however they still
+        // exist are are in-use. Neverthless, it is used to return CUDAAllocator, hence we must try to initialize it here if we can
+        // since FromProviderOptions might contain external CUDA allocator.
+        external_allocator_info = info.external_allocator_info;
+        RegisterExecutionProvider(sess, *cuda_provider_info->CreateExecutionProviderFactory(info));
+      }
+      else
+      {
+        LOGS_DEFAULT(INFO) << "CUDA execution provider is not available.";
```

Comment:
it's too dangerous to log as INFO. it should be WARNING.
a common scenario is the user may think they are running with CUDA enabled but underneath it's failing silently and everything is running on CPU.

Reply:
When the gpu pkg is installed on a CPU-only machine, logging it at WARNING might _wrongly_ indicate something is wrong.

Reply:
I didn't want it to be a warning since a CPU only user wouldn't expect a warning because cuda isn't on their system, right? The goal here is to have one version for both CPU and Cuda, I wouldn't expect a warning.

But I do think it's a problem that Python users can't explicitly say they expect the cuda provider to be used.

Reply:
> When the gpu pkg is installed on a CPU-only machine, logging it at WARNING might _wrongly_ indicate something is wrong.

But the common case is that the user has a GPU, that is why they are installing the GPU package in the first place.
so it's better to warn that it's not working than silently execute on CPU.
nobody uses INFO level by default, so it will be a silent failure  which isn't what we want.

Reply:
put another way, we shouldn't be optimizing for the case that GPU package is installed on CPU machine.
if we continue to publish both CPU and GPU packages, then we have to assume that common case for someone installing GPU package is because they installed GPU.
once we have a single CPU/GPU package, then I agree that WARNING is inappropriate.

Reply:
> I didn't want it to be a warning since a CPU only user wouldn't expect a warning because cuda isn't on their system, right? The goal here is to have one version for both CPU and Cuda, I wouldn't expect a warning.
> 
> But I do think it's a problem that Python users can't explicitly say they expect the cuda provider to be used.

yes, goal is to have a single package, but we don't have a single package currently right?

Reply:
Makes sense, I'll be sure it reports the same information the C API provides, and keep it as a warning. Then we can change it later if we ever want to remove the CPU only package.

Reply:
I removed this block since it's redundant. The C API will already log an error, this just logs an extra one that's redundant.

Reply:
Are you referring to this one? https://github.com/microsoft/onnxruntime/blob/1a5ee11dbd0dbf81977e222d243a720cbc8c5672/onnxruntime/core/framework/provider_bridge_ort.cc#L998